src/cargo/core/global_cache_tracker.rs

```diff
@@ -0,0 +1,1754 @@
+
+/// Tracking for the global shared cache (registry files, etc.).
+#[derive(Debug)]
+pub struct GlobalCacheTracker {
+    /// Connection to the SQLite database.
+    conn: Connection,
+    auto_gc_checked_this_session: bool,
+}
+
+/// This is a cache of modifications that will be saved to disk all at once
+/// via the [`DeferredGlobalLastUse::save`] method.
+///
+/// This is here to improve performance.
+#[derive(Debug)]
+pub struct DeferredGlobalLastUse {
+    /// Cache of registry keys, used for faster fetching.
+    ///
+    /// The key is the registry name (which is its directory name) and the
+    /// value is the `id` in the `registry_index` table.
+    registry_keys: HashMap<InternedString, i64>,
+    /// Cache of git keys, used for faster fetching.
+    ///
+    /// The key is the git db name (which is its directory name) and the value
+    /// is the `id` in the `git_db` table.
+    git_keys: HashMap<InternedString, i64>,
+
+    /// New registry index entries to insert.
+    registry_index_timestamps: HashMap<RegistryIndex, Timestamp>,
+    /// New registry `.crate` entries to insert.
+    registry_crate_timestamps: HashMap<RegistryCrate, Timestamp>,
+    /// New registry src directory entries to insert.
+    registry_src_timestamps: HashMap<RegistrySrc, Timestamp>,
+    /// New git db entries to insert.
+    git_db_timestamps: HashMap<GitDb, Timestamp>,
+    /// New git checkout entries to insert.
+    git_checkout_timestamps: HashMap<GitCheckout, Timestamp>,
+    /// This is used so that a warning about failing to update the database is
+    /// only displayed once.
+    save_err_has_warned: bool,
+    /// The current time, used to improve performance to avoid accessing the
+    /// clock hundreds of times.
+    now: Timestamp,
+}
+
+/// The key for a registry index entry stored in the database.
+#[derive(Clone, Debug, Hash, Eq, PartialEq)]
+pub struct RegistryIndex {
+    pub encoded_registry_name: InternedString,
+}
+
+/// The key for a registry `.crate` entry stored in the database.
+#[derive(Clone, Debug, Hash, Eq, PartialEq)]
+pub struct RegistryCrate {
+    pub encoded_registry_name: InternedString,
+    pub crate_filename: InternedString,
+    pub size: u64,
+}
+
+/// The key for a registry src directory entry stored in the database.
+#[derive(Clone, Debug, Hash, Eq, PartialEq)]
+pub struct RegistrySrc {
+    pub encoded_registry_name: InternedString,
+    pub package_dir: InternedString,
+    // Total size of the src directory in bytes.
+    //
+    // This can be None when the size is unknown. For example, when the src
+    // directory already exists on disk, and we just want to update the
+    // last-use timestamp. We don't want to take the expense of computing disk
+    // usage unless necessary. `populate_untracked_src` will handle any actual
+    // NULL values in the database, which can happen when the src directory is
+    // created by an older version of cargo that did not track sizes.
+    pub size: Option<u64>,
```

Comment:
If we still wanted to scale this PR back further, I think cutting out disk size based GC operations would be a big help as it would reduce the du code.  It'd also remove the du runtime overhead so we can better evaluate what that parts impact is to overall cache tracking performance.